workflow/sync/sync_manager.go

```diff
@@ -113,102 +111,146 @@ func (cm *Manager) Initialize(wfs []wfv1.Workflow) {
 }
 
 // TryAcquire tries to acquire the lock from semaphore.
-// It returns status of acquiring a lock , status of Workflow status updated, waiting message if lock is not available and any error encountered
-func (cm *Manager) TryAcquire(wf *wfv1.Workflow, nodeName string, syncLockRef *wfv1.Synchronization) (bool, bool, string, error) {
-	cm.lock.Lock()
-	defer cm.lock.Unlock()
+// It returns status of acquiring a lock , status of Workflow status updated, waiting message if lock is not available, the failed lock, and any error encountered
+func (sm *Manager) TryAcquire(wf *wfv1.Workflow, nodeName string, syncLockRef *wfv1.Synchronization) (bool, bool, string, string, error) {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
 
 	if syncLockRef == nil {
-		return false, false, "", fmt.Errorf("cannot acquire lock from nil Synchronization")
+		return false, false, "", "", fmt.Errorf("cannot acquire lock from nil Synchronization")
 	}
 
-	syncLockName, err := GetLockName(syncLockRef, wf.Namespace)
+	failedLockName := ""
+	syncItems, err := allSyncItems(syncLockRef)
 	if err != nil {
-		return false, false, "", fmt.Errorf("requested configuration is invalid: %w", err)
+		return false, false, "", failedLockName, fmt.Errorf("requested configuration is invalid: %w", err)
 	}
+	holderKey := getHolderKey(wf, nodeName)
 
-	lockKey := syncLockName.EncodeName()
-	lock, found := cm.syncLockMap[lockKey]
-	if !found {
-		switch syncLockRef.GetType() {
-		case wfv1.SynchronizationTypeSemaphore:
-			lock, err = cm.initializeSemaphore(lockKey)
-		case wfv1.SynchronizationTypeMutex:
-			lock = cm.initializeMutex(lockKey)
-		default:
-			return false, false, "", fmt.Errorf("unknown Synchronization Type")
-		}
+	var msg string
+	lockKeys := make([]string, len(syncItems))
+	allAcquirable := true
+	for i, syncItem := range syncItems {
+		syncLockName, err := getLockName(syncItem, wf.Namespace)
 		if err != nil {
-			return false, false, "", err
+			return false, false, "", failedLockName, fmt.Errorf("requested configuration is invalid: %w", err)
 		}
-		cm.syncLockMap[lockKey] = lock
+		lockKeys[i] = syncLockName.encodeName()
 	}
+	for i, lockKey := range lockKeys {
```

Comment:
I think Code will maintain the order of locks.
How will the code handle the below deadlock scenario?
WF1 configured 3 mutex locks (l1,l2,l3)
WF2 configured 2 locks (l3,l2)
WF2 acquired l3 and waiting for L2
WF1 acquired L1, L2 and waiting for L3

Reply:
The old code could deadlock, in that you could nest templates which performed a sequence that caused two workflows to hold and request locks that they could never escape from.
For example
* WF1 template `tp1` acquires mutex `l1`, and later `tp2` which wants mutex `l2`
* WF2 template `tp3` acquires mutex `l2` and then calls `tp1`  then it will then be deadlocked.

I don't think we should attempt to mitigate deadlocks - but providing a good UI around how locks are being held in a more accessible way would be something I'd like to have.

In your scenario I am assuming the locks are being acquired at the same time. There is not a problem:
* When WF2 runs it will only be able to either acquire **both** locks or neither.
* [`checkAcquire`](https://github.com/argoproj/argo-workflows/blob/22b64f98998754216e56bbf768fc66f92ecb0f5e/workflow/sync/sync_manager.go#L175) is called to ensure we can acquire **all** the locks we want, and then we acquire them.
* (This is done under the protection of the go mutex for sync manager).
* So either WF2 will have the two locks it needs, or WF1 will have the 3 locks it needs. Neither can hold partial locks under the multiple lock code. (They can with nested templates, but they could before).

Reply:
Yup agree with @Joibel on this one

Reply:
> providing a good UI around how locks are being held in a more accessible way would be something I'd like to have.

There is a feature request for this in #11688

> I don't think we should attempt to mitigate deadlocks

I generally agree on this, as this is basically impossible to do well and becomes a general programming bug/problem to deal with (including Turing completeness and what not).

There is a simple case that is out of scope of this PR that I believe is currently allowed but could be mitigated:
When a Workflow's own templates cause a deadlock.